From 5be77af5720e28df01531ed7c604463d9269d2ce Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Fri, 29 Aug 2014 19:01:25 -0400
Subject: [net] bridge: use list_for_each_entry_continue_reverse

Message-id: <1409338885-30414-14-git-send-email-vyasevic@redhat.com>
Patchwork-id: 89910
O-Subject: [RHEL7 PATCH 13/13] bridge: use list_for_each_entry_continue_reverse
Bugzilla: 896669
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

From: Li RongQing <roy.qing.li@gmail.com>

use list_for_each_entry_continue_reverse to rollback in fdb_add_hw
when add address failed

Signed-off-by: Li RongQing <roy.qing.li@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit a3f5ee71cdec56543e0420a788dc7df335944dd5)
Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/br_fdb.c b/net/bridge/br_fdb.c
index e57b1ca..2d1ca38 100644
--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -90,7 +90,7 @@ static void fdb_rcu_free(struct rcu_head *head)
 static void fdb_add_hw(struct net_bridge *br, const unsigned char *addr)
 {
  int err;
- struct net_bridge_port *p, *tmp;
+ struct net_bridge_port *p;
 
  ASSERT_RTNL();
 
@@ -104,11 +104,9 @@ static void fdb_add_hw(struct net_bridge *br, const unsigned char *addr)
 
  return;
 undo:
- list_for_each_entry(tmp, &br->port_list, list) {
-  if (tmp == p)
-   break;
-  if (!br_promisc_port(tmp))
-   dev_uc_del(tmp->dev, addr);
+ list_for_each_entry_continue_reverse(p, &br->port_list, list) {
+  if (!br_promisc_port(p))
+   dev_uc_del(p->dev, addr);
  }
 }
 
-- 
1.7.1